### PR TITLE
[NTP Search]: Improve handling of enabled/disabled search engines

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves https://github.com/brave/brave-browser/issues/38976
Resolves https://github.com/brave/brave-browser/issues/38977

[Screencast from 2024-06-12 10-06-04.webm](https://github.com/brave/brave-core/assets/7678024/b0ae7f04-57b6-4e06-8861-e796dbd56d42)

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

See issue

### CODE REVIEW

File: components/brave_new_tab_ui/components/search/config.ts [hunk 1]
```diff
@@ -26,7 +26,17 @@ const getConfig = () => {
   return cache!
 }
 
-export const setEngineEnabled = (engine: SearchEngineInfo, enabled: boolean) => {
+// Determines whether any search engines are enabled
+export const hasEnabledEngine = () => Object.values(getConfig()).find(c => c)
```
Comment (sangwoo108) [inline components/brave_new_tab_ui/components/search/config.ts:30]: q. could using `.length` be more readable(as long as it's what's intended)?

```suggestion
export const hasEnabledEngine = () => Object.values(getConfig()).length
```

File: components/brave_new_tab_ui/hooks/PrefHookManager.ts [hunk 2]
```diff
@@ -104,27 +104,25 @@ export class PrefHookManager<T> {
...
+            const handler: PrefListener<OnType, PrefType> = (_, newValue) => setValue(newValue)
             prefManager.addPrefListener(pref, handler)
             return () => {
                 prefManager.removePrefListener(pref, handler)
             }
         }, [prefManager, pref])
 
         return [
-            value,
+            prefManager.getPref(pref),
```
Comment (sangwoo108) [inline components/brave_new_tab_ui/hooks/PrefHookManager.ts:125]: Q. I'm curios why this change was needed.

**Review by sangwoo108 — APPROVED** [review]
LGTM

File: components/brave_new_tab_ui/components/search/config.ts (hunk 1 above)
Comment (fallaciousreasoning) [reply components/brave_new_tab_ui/components/search/config.ts:30]: It would have to be `.filter(c => c).length` as the disabled search engines are still stored (as `false`)

File: components/brave_new_tab_ui/hooks/PrefHookManager.ts (hunk 2 above)
Comment (fallaciousreasoning) [reply components/brave_new_tab_ui/hooks/PrefHookManager.ts:125]: There was a caching issue - not 100% sure what was causing it

**Comment by stephendonner** [comment]
## Verification `IN-PROGRESS` using

```
Brave | 1.69.55 Chromium: 127.0.6533.17 (Official Build) nightly (64-bit)
-- | --
Revision | c2c742fba6902889df5e08e3f3f276f5c33f048e
OS | Windows 10 Version 22H2 (Build 19045.4598)
```
